api/test/e2enew/upstream/upstream_test.go

```diff
@@ -0,0 +1,510 @@
+				 "uri": "/server_port",
+				 "upstream_id": "1"
+			 }`,
+			Headers:      map[string]string{"Authorization": base.GetToken()},
+			ExpectStatus: http.StatusOK,
+			Sleep:        base.SleepTime,
+		})
+	})
+
+	ginkgo.It("hit routes(upstream weight 1)", func() {
+		t := ginkgo.GinkgoT()
+		//hit routes
+		basepath := "http://127.0.0.1:9080/"
+		request, err := http.NewRequest("GET", basepath+"/server_port", nil)
+		assert.Nil(t, err)
+		request.Header.Add("Authorization", base.GetToken())
+		res := map[string]int{}
+		for i := 0; i < 18; i++ {
+			resp, err := http.DefaultClient.Do(request)
+			assert.Nil(t, err)
+			respBody, err := ioutil.ReadAll(resp.Body)
+			assert.Nil(t, err)
+			body := string(respBody)
+			if _, ok := res[body]; !ok {
+				res[body] = 1
+			} else {
+				res[body]++
+			}
+			resp.Body.Close()
+		}
+		assert.Equal(t, 18, res["1982"])
+	})
+
+	ginkgo.It("create chash upstream with key (remote_addr, weight equal 0 or 1)", func() {
+		base.RunTestCase(base.HttpTestCase{
+			Object: base.ManagerApiExpect(),
+			Method: http.MethodPut,
+			Path:   "/apisix/admin/upstreams/1",
+			Body: `{
+				"nodes": [
+				{
+					"host": "172.16.238.20",
+					"port": 1980,
+					"weight": 1
+				},
+				{
+					"host": "172.16.238.20",
+					"port": 1981,
+					"weight": 0
+				},
+				{
+					"host": "172.16.238.20",
+					"port": 1982,
+					"weight": 0
+				}],
+				"type": "chash",
+				"hash_on":"header",
+				"key": "remote_addr"
+			}`,
+			Headers:      map[string]string{"Authorization": base.GetToken()},
+			ExpectStatus: http.StatusOK,
+		})
+	})
+
+	ginkgo.It("create route using the upstream just created", func() {
+		base.RunTestCase(base.HttpTestCase{
+			Object: base.ManagerApiExpect(),
+			Method: http.MethodPut,
+			Path:   "/apisix/admin/routes/1",
+			Body: `{
+				"uri": "/server_port",
+				"upstream_id": "1"
+			}`,
+			Headers:      map[string]string{"Authorization": base.GetToken()},
+			ExpectStatus: http.StatusOK,
+			Sleep:        base.SleepTime,
+		})
+	})
+
+	ginkgo.It("hit routes(remote_addr, weight equal 0 or 1)", func() {
+		t := ginkgo.GinkgoT()
+		basepath := "http://127.0.0.1:9080/"
+		request, err := http.NewRequest("GET", basepath+"/server_port", nil)
+		assert.Nil(t, err)
+		request.Header.Add("Authorization", base.GetToken())
+		count := 0
+		for i := 0; i <= 17; i++ {
+			resp, err := http.DefaultClient.Do(request)
+			assert.Nil(t, err)
+			respBody, err := ioutil.ReadAll(resp.Body)
+			if string(respBody) == "1980" {
+				count++
+			}
+			resp.Body.Close()
+		}
+		assert.Equal(t, 18, count)
+	})
+	ginkgo.It("create chash upstream with key (remote_addr, all weight equal 0)", func() {
+		base.RunTestCase(base.HttpTestCase{
+			Object: base.ManagerApiExpect(),
+			Method: http.MethodPut,
+			Path:   "/apisix/admin/upstreams/1",
+			Body: `{
+				 "nodes": [
+				 {
+					 "host": "172.16.238.20",
+					 "port": 1980,
+					 "weight": 0
+				 },
+				 {
+					 "host": "172.16.238.20",
```

Comment:
ditto
and please replace them all, thanks.

Reply:
done.